src/scoped_map.rs

```diff
@@ -0,0 +1,106 @@
+//! Scoped hashmap.
+//!
+//! This container keeps a map from a type `K` to `V`, but with a
+//! twist: it supports "push" and "pop" operations, which alter a
+//! stack-level, and on "pop", any mappings created at the current
+//! stack level or greater are removed. This is useful in conjunction
+//! with tree-structured walks over code: for example, a GVN pass that
+//! traverses the domtree, keeping "availability" of already-computed
+//! expressions in the map only with the scope of a subtree.
+//!
+//! For efficiency, the scoped hashmap does not actually perform a
+//! "remove" operation in the underlying hashmap for every operation
+//! removed by a pop. This would result in O(n) pop cost,
+//! worst-case. Instead, we fast-invalidate all mappings by
+//! incrementing a generation number, and checking that generation on
+//! access, ignoring stale entries. We need a separate generation
+//! number for each level of the scoped hashmap; "push" increments the
+//! generation on the new level ("this is the 64th time we have been
+//! at level 3") and when a value is inserted, it records the level it
+//! is at and the current generation of that level.
+//!
+//! Note that this scheme does *not* support shadowing: if we set a
+//! particular key `k1` to value `v1` at level 3, then set it again at
+//! level 4, we will remove it when we pop level 4; the old value set
+//! at level 3 does not re-appear. Doing so would require a more
+//! complex data structure, and is unnecessary for our use-cases.
```

Comment:
In particular, we don't do things like
```
if x == 0 {
   // insert a definition of `x` as 0 in this subtree
}
```
which would use this kind of shadowing.

Reply:
Indeed; added a note why we don't need shadowing in the specific context of GVN (its one use here).